src/UCTNode.cpp

```diff
@@ -135,35 +132,36 @@ void UCTNode::link_nodelist(std::atomic<int> & nodecount,
     LOCK(get_mutex(), lock);
 
     for (const auto& node : nodelist) {
-        m_children.emplace_back(
-            std::make_unique<UCTNode>(node.second, node.first, init_eval)
-        );
+        m_child_scores.push_back({node.second, node.first});
     }
 
-    nodecount += m_children.size();
+    m_child_init_eval = init_eval;
+
+    nodecount += m_child_scores.size();
     m_has_children = true;
 }
 
+// Only safe to call prior to uct_select_child().
 void UCTNode::kill_superkos(const KoState& state) {
-    for (auto& child : m_children) {
-        auto move = child->get_move();
+    assert(m_expanded.size() == 0);
+
+    std::vector<std::pair<int, float>> good_moves;
+    for (auto& child : m_child_scores) {
+        auto move = child.first;
         if (move != FastBoard::PASS) {
             KoState mystate = state;
             mystate.play_move(move);
 
             if (mystate.superko()) {
-                // Don't delete nodes for now, just mark them invalid.
-                child->invalidate();
+                // Skip invalid moves.
+                continue;
             }
         }
+        good_moves.push_back(child);
```

Comment:
Why not reorder the children inplace (`std::partition`) and use `erase` afterwards? Could save some copying and entries are only ordered in `m_expanded`.